ai_ta_backend/vector_database.py

```diff
@@ -1080,10 +1169,33 @@ def getTopContexts(self, search_query: str, course_name: str, token_limit: int =
       top_n = 80 # HARD CODE TO ENSURE WE HIT THE MAX TOKENS
       start_time_overall = time.monotonic()
 
-      found_docs: list[Document] = self.vector_search(search_query=search_query, course_name=course_name)
+      # Vector search with ONLY original query
+      # found_docs: list[Document] = self.vector_search(search_query=search_query, course_name=course_name)
+
+      # Multi query retriever
+      generate_queries = (
+                MULTI_QUERY_PROMPT
+                | self.llm
+                | StrOutputParser()
+                | (lambda x: x.split("\n"))
+                | (lambda x: list(filter(None, x)))  # filter out non-empty strings
+            )
       
-      # call context padding function here
+      generated_queries = generate_queries.invoke({"original_query": search_query})
+      print("generated_queries", generated_queries)
+
+      batch_found_docs: list[list[Document]] = self.batch_vector_search(search_queries=generated_queries, course_name=course_name)
+
+      found_docs = self.reciprocal_rank_fusion(batch_found_docs)
+      found_docs = [doc for doc, score in found_docs]
+      print(f"Number of docs found with multiple queries: {len(found_docs)}")
+      if len(found_docs) == 0:
+          return []
+
+      # 'context padding' // 'parent document retriever' 
+      # TODO maybe only do context padding for top 5 docs? Otherwise it's wasteful imo.
       final_docs = self.context_padding(found_docs, search_query, course_name)

```

Comment:
I think we should maybe just do the top 5 docs w/ extra padding. Otherwise, it feels too wasteful to fill our context window with padding for ALL of our contexts, when that space could be used by other relevant contexts.

Reply:
Yup! I just implemented parallel processing for `found_docs[5:]` (just re-factoring the objects to be like Supabase output) and its working well! I am working on doing the same for the first 5 docs now.